node/src/transaction_pool/candidate/transaction_pool_candidate_actions.rs

```diff
@@ -113,8 +113,14 @@ impl redux::EnablingCondition<crate::State> for TransactionPoolCandidateAction {
                 .is_some(),
             TransactionPoolCandidateAction::Libp2pTransactionsReceived { .. } => true,
             TransactionPoolCandidateAction::VerifyNext => {
-                // TODO: if a block is being applied or produced, skip this action too
-                state.transition_frontier.sync.is_synced()
+                // Don't continue if we are producing a block, or we never synced yet
+                // or if the ledger service is busy.
+                !state.block_producer.is_producing()
+                    && state
+                        .transition_frontier
+                        .best_tip()
+                        .is_some_and(|b| !b.is_genesis())
+                    && !state.ledger.write.is_busy()
```

Comment:
@binier what do you think of this version? maybe I should add a similar check for reads? (I remember there was some prioritization going on but I don't remember the details, not even sure if we have any heavy reads that would delay anything)

Reply:
> Sounds good except is_synced is not really necessary (for a brief moment at the beginning the node will consider himself in sync?)

At the beginning node will not be synced. Then it will reconstruct genesis ledger/block, make it our best tip and go to synced state.

with above solution, it will not include any txs after genesis block (given that we are starting up a new network or doing hard fork). That might be a reasonable tradeoff though, so maybe it's fine.

Anyways, I think above solution is decent enough.

> maybe I should add a similar check for reads?

No, we could have reads all the time. E.g. peer requesting parts of the ledger.